### PR TITLE
new resource and data source "azurerm_key_vault_managed_hardware_security_module"

### PR DESCRIPTION
**Pull request by njuCZ**
first pr to fix issue: #10209

1. this resource is only supported in latest preview version: 2020-04-01-preview, and it's going to be GA soon
2. this resource is almost the same with key vault, doc refers to here: https://docs.microsoft.com/en-us/azure/key-vault/managed-hsm/overview
3. api issue that needs to pay attention:
3.1 `soft_delete` is always `true`. If setting to `false` in the create rest api, it will report error
3.2 `soft_delete_retention_days` could not be updated, the backend service will report error
3.3 `purge_protection_enabled` could be set in the update rest api, but the response remain the same. this is an issue and I have reported to the service team
3.4 I have made the above fields `forcenew`
4. During public preview, the rest api limits we can create only 1 HSM pool(s) per subscription in a region.  The acctest is made to  run in sequence.

image

### CODE REVIEW

File: azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource_test.go:113]: rather than inlining this function call - which makes it harder to trace these through, can we split this out into a variable:

```suggestion
`, template, data.RandomInteger)
```

File: azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource_test.go:145]: (same here)

File: azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource_test.go:164]: (same here)

File: azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource_test.go:171]: the provider block shouldn't be in the template, but instead in each test, since this could vary per test - can we fix this?

File: azurerm/internal/services/keyvault/resourceids.go
Comment (tombuildsstuff) [inline azurerm/internal/services/keyvault/resourceids.go:4]: this is a Managed hardware security module, so presumably this can become:

```suggestion
//go:generate go run ../../tools/generator-resource-id/main.go -path=./ -name=ManagedHSM -id=/subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resGroup1/providers/Microsoft.KeyVault/managedHSMs/hsm1
```

which also fixes the naming to be `Name` rather than `H S M`

File: website/docs/d/key_vault_managed_hardware_security_module.html.markdown
Comment (tombuildsstuff) [inline website/docs/d/key_vault_managed_hardware_security_module.html.markdown:30]:
```suggestion
* `name` - The name of the Key Vault Managed Hardware Security Module.
```

File: website/docs/r/key_vault_managed_hardware_security_module.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/key_vault_managed_hardware_security_module.html.markdown:58]: this isn't defaulted in the schema - and from memory Terraform intentionally defaults to `7` for the other resource, so we should match that here

File: website/docs/r/key_vault_managed_hardware_security_module.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/key_vault_managed_hardware_security_module.html.markdown:54]:
```suggestion
* `tenant_id` - (Required) The Azure Active Directory Tenant ID that should be used for authenticating requests to the key vault Managed Hardware Security Module.
```

File: website/docs/r/key_vault_managed_hardware_security_module.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/key_vault_managed_hardware_security_module.html.markdown:52]: can we sort this list alphabetically?

File: website/docs/r/key_vault_managed_hardware_security_module.html.markdown
Comment (njuCZ) [reply website/docs/r/key_vault_managed_hardware_security_module.html.markdown:58]: for the key vault resource, it's also default 90 https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/keyvault/key_vault_resource.go#L196

File: azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource..go
Comment (njuCZ) [reply azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource..go:221]: yes, after testing, this is a service team's issue. I will submit an issue soon and email them to give feedback

**Comment by njuCZ** [comment]
it seems the CI error is because of temporary network break

**Comment by njuCZ** [comment]
rebase and run the acctest in sequence

File: azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource..go
Comment (katbyte) [reply azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource..go:221]: could we link to that in code please?

**Review by katbyte — CHANGES_REQUESTED** [review]

File: azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource..go
Comment (katbyte) [inline azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource..go:221]:
```suggestion
	// there is an API bug being tracked here: https://github.com/Azure/azure-rest-api-specs/issues/13365
	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
		if response.WasNotFound(future.Response()) {
			return nil
		}
```

**Comment by njuCZ** [comment]
thanks for the commit

**Comment by njuCZ** [comment]
@katbyte I have made the acctest run in sequence. Every test has used the func `data.ResourceSequentialTest`.

**Comment by njuCZ** [comment]
@katbyte thanks for the suggestions. I have changed the way to run test case. Now all acctest have passed in teamcity

**Review by katbyte — APPROVED** [review]
Thanks @njuCZ - LGTM and once the merge conflict is resolved we can get this merged

**Comment by njuCZ** [comment]
hi, @katbyte I have fixed the conflict. In the master branch, file `key_vault_migration_resource.go` is deleted. so I just add a commit to delete the files.

**Comment by njuCZ** [comment]
@katbyte Have rebased and fix the conflicts

**Pull request merged**
